src/main/java/com/linecorp/armeria/client/Clients.java

```diff
@@ -131,6 +135,51 @@
                                      .build(interfaceClass);
     }
 
+    /**
+     * Creates a new derived client that connects to the same {@link URI} with the specified {@code client}
+     * with the specified {@code additionalOptions}. Note that the derived client will use the options of
+     * the specified {@code client} unless specified in {@code additionalOptions}.
+     */
+    public static <T> T newDerivedClient(T client, ClientOptionValue<?>... additionalOptions) {
+        requireNonNull(additionalOptions, "additionalOptions");
+        // TODO(trustin): Reduce the cost of creating a derived ClientOptions.
+        return newDerivedClient(client, baseOptions -> ClientOptions.of(baseOptions, additionalOptions));
+    }
+
+    /**
+     * Creates a new derived client that connects to the same {@link URI} with the specified {@code client}
+     * with the specified {@code additionalOptions}. Note that the derived client will use the options of
+     * the specified {@code client} unless specified in {@code additionalOptions}.
+     */
+    public static <T> T newDerivedClient(T client, Iterable<ClientOptionValue<?>> additionalOptions) {
+        requireNonNull(additionalOptions, "additionalOptions");
+        // TODO(trustin): Reduce the cost of creating a derived ClientOptions.
+        return newDerivedClient(client, baseOptions -> ClientOptions.of(baseOptions, additionalOptions));
+    }
+
+    private static <T> T newDerivedClient(T client,
+                                          Function<ClientOptions, ClientOptions> optionFactory) {
+
+        requireNonNull(client, "client");
+        if (!(client instanceof ProxiedClient)) {
+            throw new IllegalArgumentException("not a client: " + client);
+        }
+
+        final ClientInvocationHandler parent =
```

Comment:
I think we can avoid the weird name mangling with something like

```
ProxiedClient.getClass().getMethod("handler").invoke(client);
```

The method.getDeclaredClass() == ProxiedClient.class check in the handler will be true for this call and should be false for any other handler() method.

Since this is a reflective Java proxy, I don't think the performance will be any different for the .invoke() call vs this invocation.